server/storage/azure.go

```diff
@@ -0,0 +1,129 @@
+	logger          *log.Logger
+}
+
+func getCredentials() (*azidentity.DefaultAzureCredential, error) {
+	return azidentity.NewDefaultAzureCredential(nil)
+}
+
+func NewAzureBlobStorage(ctx context.Context, storageAccountName string, containerName string, logger *log.Logger) (Storage, error) {
+	credentials, err := getCredentials()
+	if err != nil {
+		return nil, err
+	}
+
+	serviceUrl := fmt.Sprintf("https://%s.blob.core.windows.net", storageAccountName)
+	client, err := azblob.NewClient(serviceUrl, credentials, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	containerClient := client.ServiceClient().NewContainerClient(containerName)
+
+	azureStorage := &AzureStorage{
+		client:          client,
+		containerClient: containerClient,
+		containerName:   containerName,
+		logger:          logger,
+	}
+	return azureStorage, nil
+}
+
+func (s *AzureStorage) Type() string {
+	return "azure"
+}
+
+func (s *AzureStorage) Get(ctx context.Context, token string, filename string, rng *Range) (io.ReadCloser, uint64, error) {
+	key := fmt.Sprintf("%s/%s", token, filename)
+	blobClient := s.containerClient.NewBlobClient(key)
+
+	var options *azblob.DownloadStreamOptions
+	if rng != nil {
+		options = &azblob.DownloadStreamOptions{
+			Range: azblob.HTTPRange{
+				Offset: int64(rng.Start),
+				Count:  int64(rng.Limit),
+			},
+		}
+	}
+
+	resp, err := blobClient.DownloadStream(ctx, options)
+	if err != nil {
+		return nil, 0, err
+	}
+	return resp.Body, uint64(*resp.ContentLength), nil
+}
+
+func (s *AzureStorage) Head(ctx context.Context, token string, filename string) (contentLength uint64, err error) {
+	key := fmt.Sprintf("%s/%s", token, filename)
+	props, err := s.containerClient.NewBlobClient(key).GetProperties(ctx, nil)
+	if err != nil {
+		return 0, err
+	}
+	return uint64(*props.ContentLength), nil
+}
+
+func (s *AzureStorage) Put(ctx context.Context, token string, filename string, reader io.Reader, _ string, _ uint64) error {
+	key := fmt.Sprintf("%s/%s", token, filename)
+	_, err := s.client.UploadStream(ctx, s.containerName, key, reader, nil)
+	return err
+}
+
+func (s *AzureStorage) Delete(ctx context.Context, token string, filename string) error {
+	key := fmt.Sprintf("%s/%s", token, filename)
+	blobClient := s.containerClient.NewBlobClient(key)
+	_, err := blobClient.Delete(ctx, nil)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *AzureStorage) IsNotExist(err error) bool {
+	return bloberror.HasCode(err, bloberror.BlobNotFound)
+}
+
+func (s *AzureStorage) Purge(ctx context.Context, days time.Duration) error {
+	pager := s.containerClient.NewListBlobsFlatPager(nil)
+
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return err
+		}
+
+		for _, blob := range page.Segment.BlobItems {
+			if time.Since(*blob.Properties.LastModified) > days {
+				blobClient := s.containerClient.NewBlobClient(*blob.Name)
+				_, err := blobClient.Delete(ctx, nil)
+				if err != nil {
+					continue
```

Comment:
i guess you want to do `return err` here.
continue doesn't do anything, since there are no extra instructions after the if, and `break`would loose the `err` and `return nil` after the whole loop is over

is there a way to initiate a "general `blobClient`" outside of the loop and just override the `blob.Name`, by chance?

Reply:
> is there a way to initiate a "general `blobClient`" outside of the loop and just override the `blob.Name`, by chance?

https://github.com/Azure/azure-sdk-for-go/blob/90c29ccd6952363d37f7f2c2aa2e8d329ee59015/sdk/storage/azblob/client.go#L114-L116

i guess in general you can get rid of the `AzureStorage.containerClient` and `AzureStorage.blobClient` and use the helpers on that package anywhere.

Reply:
Thank you for all the feedback.  I will revisit soon, apologies I couldn't get to it this weekend :)

Reply:
Good idea, made that change to not create clients manually.